func-puzzles.js

```diff
@@ -0,0 +1,177 @@
+let funcLib = {};
+
+arrLib = (function (context) {
+    context.partial = function (g, ...args1) {
+        return function (...args2) {
+            return g(...args1, ...args2);
+        }
+    };
+
+    context.curry = function (f) {
+        return function currify() {
+            const args = Array.prototype.slice.call(arguments);
+            return args.length >= f.length ?
+                f.apply(null, args) :
+                currify.bind(null, ...args)
+        }
+    };
+
+    context.linFold = function (array, callback, initialValue = 0) {
+        let accum = initialValue;
+        for (let i = 0; i < array.length; i++) {
+            accum = callback(accum, array[i], i, initialValue);
+        }
+        return accum;
+    };
+
+    context.map = function (array, callback) {
+        let res = [];
+        for (let item of array) {
+            res.push(callback(item));
+        }
+        return res;
+    };
+
+    context.filter = function (array, callback) {
+        let res = [];
+        for (let item of array) {
+            if (callback(item)) {
+                res.push(item);
+            }
+        }
+        return res;
+    };
+
+    context.avgEvenNums = function (array) {
+        let evens = context.filter(array, x => x % 2 === 0);
+        let sum = context.linFold(evens, (accum, item, index, initialVal) => {
+            return accum + item
+        });
+        return sum / evens.length;
+    };
+
+    context.lazy = function (f) {
+        return function (...args) {
```

Comment:
What you forgot is to store (memoize) the result.
Here u can find an example and also the problem of lazy functions: https://www.codementor.io/@agustinchiappeberrini/lazy-evaluation-and-javascript-a5m7g8gs3